File: tree-medium/98.验证二叉搜索树.java
/*
 * @lc app=leetcode.cn id=98 lang=java
 *
 * [98] 验证二叉搜索树
 */

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

// @lc code=start

// 77/77 cases passed (0 ms)
// Your runtime beats 100 % of java submissions
// Your memory usage beats 12.05 % of java submissions (38.3 MB)
class Solution {
    public boolean isValidBST(TreeNode root) {
        return isValid(root, null, null);
    }

    boolean isValid(TreeNode root, TreeNode min, TreeNode max) {
        if (root == null) {
            return true;
        }
        if (min != null && root.val <= min.val) {
            return false;
        }
        if (max != null && root.val >= max.val) {
            return false;
        }
        return isValid(root.left, min, root) && isValid(root.right, root, max);
    }
}
// @lc code=end

File: tree-medium/98.验证二叉搜索树.py
#
# @lc app=leetcode.cn id=98 lang=python3
#
# [98] 验证二叉搜索树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
# 52ms 15.8MB# 52ms 15.8MB# 52ms 15.8MB# 52ms 15.8MB# 52ms 15.8MB# 52ms 15.8MB# 52ms 15.8MB# 52ms 15.8MB# 52ms 15.8MB# 52ms 15.8MB# 52ms 15.8MB# 52ms 15.8MB


# 52ms 15.8MB
# @BFS
class Solution:
    def isValidBST(self, root):
        def BFS(root, left, right):
            if root is None:
                return True
            if left < root.val < right:
                return BFS(root.left, left, root.val) and BFS(root.right, root.val, right)
            else:
                return False

        return BFS(root, -float("inf"), float("inf"))


# @lc code=end
